File: Windows Source/TotallyHuman/ML/LogisticRegression.cs
using System;
using System.Collections.Generic;
using System.Linq;
using TotallyHuman.Models;

namespace TotallyHuman.ML;

/// <summary>
/// Logistischer Klassifikator mit Gradient Descent und L2-Regularisierung.
/// 1:1 Port von LogisticRegression.swift (learningRate=0.05, iterations=1500,
/// l2Strength=0.001, tolerance=1e-6).
/// </summary>
public sealed class LogisticRegression
{
    public struct Configuration
    {
        public double LearningRate;
        public int Iterations;
        public double L2Strength;
        public double Tolerance;

        public static Configuration Default => new()
        {
            LearningRate = 0.05,
            Iterations = 1500,
            L2Strength = 0.001,
            Tolerance = 1e-6
        };
    }

    public LogistischesModell Model { get; private set; }
    public Configuration Config;

    public LogisticRegression(LogistischesModell? model = null,
                              int featureCount = AnalysisConfig.FeatureVectorSize,
                              Configuration? configuration = null)
    {
        Config = configuration ?? Configuration.Default;
        if (model != null)
        {
            Model = model;
        }
        else
        {
            Model = new LogistischesModell
            {
                Gewichte = new double[featureCount],
                Bias = 0.0,
                TrainingsZeit = DateTime.Now,
                AnzahlBeispiele = 0,
                KreuzvalidierungsGenauigkeit = 0.0
            };
        }
    }

    /// <summary>Trainiert das Modell anhand von Proben.</summary>
    public LogistischesModell Train(IReadOnlyList<TrainingsProbe> samples)
    {
        if (samples.Count == 0) return Model;
        int featureCount = samples[0].Merkmalsvektor.Length;
        var weights = new double[featureCount];
        double bias = 0.0;
        double n = samples.Count;
        double previousLoss = double.MaxValue;

        for (int iter = 0; iter < Config.Iterations; iter++)
        {
            var gradientW = new double[featureCount];
            double gradientB = 0.0;
            double loss = 0.0;

            foreach (var sample in samples)
            {
                if (sample.Merkmalsvektor.Length != featureCount) continue;
                double z = Dot(weights, sample.Merkmalsvektor) + bias;
                double prediction = Sigmoid(z);
                double y = (int)sample.Label;
                double error = prediction - y;
                loss += BinaryCrossEntropy(prediction, y);
                for (int i = 0; i < featureCount; i++)
                    gradientW[i] += error * sample.Merkmalsvektor[i];
                gradientB += error;
            }

            for (int i = 0; i < featureCount; i++)
            {
                gradientW[i] = gradientW[i] / n + Config.L2Strength * weights[i];
                weights[i] -= Config.LearningRate * gradientW[i];
            }
            bias -= Config.LearningRate * (gradientB / n);

            double regularization = 0.5 * Config.L2Strength * Dot(weights, weights);
            double totalLoss = loss / n + regularization;
            if (Math.Abs(previousLoss - totalLoss) < Config.Tolerance) break;
            previousLoss = totalLoss;
        }

        Model = new LogistischesModell
        {
            Gewichte = weights,
            Bias = bias,
            TrainingsZeit = DateTime.Now,
            AnzahlBeispiele = samples.Count,
            KreuzvalidierungsGenauigkeit = EvaluateAccuracy(samples, weights, bias)
        };
        return Model;
    }

    public double PredictProbability(double[] features)
    {
        if (features.Length != Model.Gewichte.Length) return 0.5;
        return Model.Vorhersage(features);
    }

    public TrainingsLabel Predict(double[] features, double threshold = 0.5)
        => PredictProbability(features) >= threshold ? TrainingsLabel.Ki : TrainingsLabel.Echt;

    // MARK: - Intern

    private static double EvaluateAccuracy(IReadOnlyList<TrainingsProbe> samples, double[] weights, double bias)
    {
        if (samples.Count == 0) return 0.0;
        int correct = 0;
        foreach (var sample in samples)
        {
            if (sample.Merkmalsvektor.Length != weights.Length) continue;
            double p = Sigmoid(Dot(weights, sample.Merkmalsvektor) + bias);
            int predicted = p >= 0.5 ? 1 : 0;
            if (predicted == (int)sample.Label) correct++;
        }
        return (double)correct / samples.Count;
    }

    private static double BinaryCrossEntropy(double prediction, double label)
    {
        double p = Math.Min(Math.Max(prediction, 1e-12), 1.0 - 1e-12);
        return -(label * Math.Log(p) + (1.0 - label) * Math.Log(1.0 - p));
    }

    private static double Sigmoid(double x)
    {
        if (x >= 0) return 1.0 / (1.0 + Math.Exp(-x));
        double z = Math.Exp(x);
        return z / (1.0 + z);
    }

    private static double Dot(double[] a, double[] b)
    {
        double s = 0.0;
        int len = Math.Min(a.Length, b.Length);
        for (int i = 0; i < len; i++) s += a[i] * b[i];
        return s;
    }
}

File: macOS Source/TotallyHuman/ML/LogisticRegression.swift
import Foundation
import Accelerate

/// Ein einfacher logistischer Klassifikator mit Gradient Descent und L2-Regularisierung.
final class LogisticRegression {
    struct Configuration {
        var learningRate: Double
        var iterations: Int
        var l2Strength: Double
        var tolerance: Double

        init(learningRate: Double = 0.05, iterations: Int = 1_500, l2Strength: Double = 0.001, tolerance: Double = 1e-6) {
            self.learningRate = learningRate
            self.iterations = iterations
            self.l2Strength = l2Strength
            self.tolerance = tolerance
        }
    }

    private(set) var model: LogistischesModell
    var configuration: Configuration

    init(model: LogistischesModell? = nil, featureCount: Int = AnalysisConfig.featureVectorSize, configuration: Configuration = Configuration()) {
        if let model {
            self.model = model
        } else {
            self.model = LogistischesModell(
                gewichte: Array(repeating: 0.0, count: featureCount),
                bias: 0.0,
                trainingsZeit: Date(),
                anzahlBeispiele: 0,
                kreuzvalidierungsGenauigkeit: 0.0
            )
        }
        self.configuration = configuration
    }

    /// Trainiert das Modell anhand von Proben.
    /// - Returns: Die trainierte Modellinstanz.
    @discardableResult
    func train(samples: [TrainingsProbe]) -> LogistischesModell {
        guard !samples.isEmpty else { return model }
        let featureCount = samples[0].merkmalsvektor.count
        var weights = Array(repeating: 0.0, count: featureCount)
        var bias = 0.0
        let n = Double(samples.count)
        var previousLoss = Double.greatestFiniteMagnitude

        for _ in 0..<configuration.iterations {
            var gradientW = Array(repeating: 0.0, count: featureCount)
            var gradientB = 0.0
            var loss = 0.0

            for sample in samples where sample.merkmalsvektor.count == featureCount {
                let z = dot(weights, sample.merkmalsvektor) + bias
                let prediction = sigmoid(z)
                let y = Double(sample.label.rawValue)
                let error = prediction - y
                loss += binaryCrossEntropy(prediction: prediction, label: y)
                for i in 0..<featureCount { gradientW[i] += error * sample.merkmalsvektor[i] }
                gradientB += error
            }

            for i in 0..<featureCount {
                gradientW[i] = gradientW[i] / n + configuration.l2Strength * weights[i]
                weights[i] -= configuration.learningRate * gradientW[i]
            }
            bias -= configuration.learningRate * (gradientB / n)

            let regularization = 0.5 * configuration.l2Strength * dot(weights, weights)
            let totalLoss = loss / n + regularization
            if abs(previousLoss - totalLoss) < configuration.tolerance { break }
            previousLoss = totalLoss
        }

        model = LogistischesModell(
            gewichte: weights,
            bias: bias,
            trainingsZeit: Date(),
            anzahlBeispiele: samples.count,
            kreuzvalidierungsGenauigkeit: evaluateAccuracy(samples: samples, weights: weights, bias: bias)
        )
        return model
    }

    /// Gibt eine Wahrscheinlichkeit für das Label „KI-Musik“ zurück.
    func predictProbability(features: [Double]) -> Double {
        guard features.count == model.gewichte.count else { return 0.5 }
        return model.vorhersage(merkmale: features)
    }

    /// Klassifiziert anhand eines Schwellwerts.
    func predict(features: [Double], threshold: Double = 0.5) -> TrainingsProbe.TrainingsLabel {
        predictProbability(features: features) >= threshold ? .ki : .echt
    }

    func update(with samples: [TrainingsProbe]) {
        _ = train(samples: samples)
    }

    // MARK: - Intern

    private func evaluateAccuracy(samples: [TrainingsProbe], weights: [Double], bias: Double) -> Double {
        guard !samples.isEmpty else { return 0.0 }
        var correct = 0
        for sample in samples where sample.merkmalsvektor.count == weights.count {
            let p = sigmoid(dot(weights, sample.merkmalsvektor) + bias)
            let predicted = p >= 0.5 ? 1 : 0
            if predicted == sample.label.rawValue { correct += 1 }
        }
        return Double(correct) / Double(samples.count)
    }

    private func binaryCrossEntropy(prediction: Double, label: Double) -> Double {
        let p = min(max(prediction, 1e-12), 1.0 - 1e-12)
        return -(label * log(p) + (1.0 - label) * log(1.0 - p))
    }

    private func sigmoid(_ x: Double) -> Double {
        if x >= 0 { return 1.0 / (1.0 + exp(-x)) }
        let z = exp(x)
        return z / (1.0 + z)
    }

    private func dot(_ a: [Double], _ b: [Double]) -> Double {
        zip(a, b).reduce(0.0) { $0 + $1.0 * $1.1 }
    }
}
